### ISSUE TITLE
Proposition: assumptions for functions

### ISSUE THREAD

**Issue by pelegm** [sympy/sympy#8760]
Continuing my line of thought from #8205, #8759, and with relation to the discussion at #2671, I propose here a list of "assumptions" of functions, that can be incorporated into the (old) assumption mechanism, or lie next to it.
```
additive
    preserves the addition operation
algebraic
    a function that can be defined as the root of a polynomial equation
analytic
    can be defined locally by a convergent power series
arithmetic
    a function from the positive integers into the complex numbers
bijective
    ...
concave
    ...
constant
    a function whose (output) value is the same for every input value
continuous
    ...
convex
    ...
decreasing
    ...
differentiable
    a function whose derivative exists at each point in its domain
entire
    ...
even
    is symmetric with respect to the Y-axis
exponential
    ...
holomorphic
    ...
homeomorphism
    ...
hyperbolic
    ...
identity
    ...
increasing
    ...
indicator
    a function defined on a set X that indicates membership of an element in a subset A of X, having the value 1 for all elements of A and the value 0 for all elements of X not in A
injective
    a function that preserves distinctness: it never maps distinct elements of its domain to the same element of its codomain
integrable
    ...
linear
    a polynomial function of degree zero or one, or is the zero polynomial
logarithmic
    ...
meromorphic
    ...
monotone
    ...
multiplicative
    preserves the multiplication operation
nowhere_continuous
    ...
odd
    is symmetric with respect to the origin
periodic
    a function that repeats its values in regular intervals or periods
polynomial
    an expression consisting of variables (or indeterminates) and coefficients, that involves only the operations of addition, subtraction, multiplication, and non-negative integer exponents
quadratic
    a polynomial of degree 2
rational
    a function which can be defined by a rational fraction such that both the numerator and the denominator are polynomials
smooth
    has derivatives of all orders
step
    that can be written as a finite linear combination of indicator functions of intervals
strictly_concave
    ...
strictly_convex
    ...
strictly_decreasing
    ...
strictly_increasing
    ...
strictly_monotone
    ...
subadditive
    for which the value of f(x+y) is less than or equal to f(x)+f(y)
superadditive
    for which the value of f(x+y) is greater than or equal to f(x)+f(y)
surjective
    ...
transcendental
    a function that is not algebraic
trigonometric
    ...
```

Then, just like in `core/assumptions.py`, `FactRules` can be defined and applied. What do you think?

**Comment by skirpichev** [comment]
> What do you think?

For example, I don't think how this can work with the current assumptions.  In some of your "assumptions" it the argument index of the function does matter.  E.g. `f=sin(x)*cos(y)` is even wrt `y`.

**Comment by asmeurer** [comment]
I like the idea, but I'd much rather do this in the framework of the new assumptions.

**Comment by pelegm** [comment]
@skirpichev, good point. In @asmeurer's comments in #2671, he adds `var` argument to the `is_...` relevant methods, so perhaps that's the reason for it. Indeed, you must provide the relevant argument, even if you ask for `odd` or `bounded`. But your point is clear: that's probably can't be incorporated into the existing system of (old) assumptions.

@asmeurer, you know better than I am, so I'd follow your advice about the new assumptions. I still need to understand better how these work, though. I just throw this here as a general idea, nothing practical really.

**Comment by asmeurer** [comment]
The new assumptions would let you create a predicate like `Q.even_function` that can take arguments, like `Q.even_function(f(x, y), x)` means that `f(x, y)` is even with respect to `x`. 

As far as how to write down facts for these things, that's something that's still being worked out. There are currently handlers, and also my system at https://github.com/sympy/sympy/pull/2508.

**Comment by pelegm** [comment]
@asmeurer, thanks for the explanation.

I think that I'll read #2508 more carefully, then play with new assumptions a little bit; in the meanwhile, I'll keep this issue open and fill it with "facts" that can be converted into "handlers" for new assumptions.

**Comment by pelegm** [comment]
This is a first attempt to discuss "known facts" of functional assumptions in the context of new assumptions. As a proof of concept, I've chosen the first assumption, `additive`. So these are the additive known facts; I ignore the variable issue for now, and I always assume that the function is of some symbol `x`. I'm not sure how to handle this more precisely at the moment.

Note: there are two concepts defined by the term `additive`. I am considering the algebraic definition: a function is said to be *additive* if for every two complex numbers `x` and `y`, it follows that `f(x + y) = f(x) + f(y)`. It is *subadditive* if for every `x`, `y`, it holds that `f(x + y) <= f(x) + f(y)`, and it is *superadditive* if for every `x`, `y` it holds that `f(x + y) >= f(x) + f(y)`. So the known facts are:
```python
Equivalent(Q.subadditive & Q.superadditive, Q.additive),

# Any linear map is additive
Implies(Q.linear, Q.additive),
```

The second "known facts" lists is for the second functional assumption: `algebraic`.
```python
Equivalent(Q.algebraic, ~Q.transcendental),

# Algebraic function types
Implies(Q.rational, Q.algebraic),
Implies(Q.polynomial, Q.rational),
Implies(Q.linear, Q.polynomial),
Implies(Q.constant, Q.linear),

# Transcendental function types
Implies(Q.exponential, Q.transcendental),
Implies(Q.hyperbolic, Q.transcendental),
Implies(Q.logarithmic, Q.transcendental),
Implies(Q.trigonometric, Q.transcendental),
```

Finally, here are some sporadic examples of non-trivial implications:
```
Implies(Q.linear & Q.bounded, Q.constant),
Implies(Q.linear & Q.periodic, Q.constant),
Implies(Q.periodic & Q.continuous, Q.bounded),
```

(I have put really little thought on all of the above. This is of course should be treated very carefully, and with a lot of thought. This is only some proof of concept.)

**Comment by pelegm** [comment]
And this is a proof of concept for the `AskLinearHandler(FunctionalHandler)`:
```python
def Add(expr, assumptions):
    """ Linear if all arguments are linear, unknown otherwise. """
    for arg in expr.args:
        if ask(Q.linear(arg), assumptions) in (False, None):
            return
        return True
```

This is in general quite difficult, but I'm writing it here to hear your opinion. For example, where and how can we take into account the relevant symbol/variable...

**Comment by asmeurer** [comment]
A more common term for your "additive" is "linear".

**Comment by asmeurer** [comment]
Granted the two are not entirely identical, but additive but not linear is not common.

**Comment by asmeurer** [comment]
Another issue here: assumptions on the function may be dependent on an assumption about the domain of the function. How do we represent this. For instance, sin(x) is bounded if x is real.

**Comment by asmeurer** [comment]
I like the direction this is going. One thing I would caution about is to be *very* precise in your definitions of the assumptions. Ideally each assumption should have a precise unambiguous mathematical definition. There should be no controversy over any fact that you write down. I started doing this for the existing assumptions at https://github.com/sympy/sympy/pull/2538, but it wasn't finished (partly because the actual definitions aren't decided upon yet and partly because I didn't have time---feel free to continue work off of any of my pull requests).

**Comment by pelegm** [comment]
> assumptions on the function may be dependent on an assumption about the domain of the function. How do we represent this. For instance, sin(x) is bounded if x is real.

No argue this is complicated. However, I can imagine how to represent that in a system which resembles the old assumptions. Perhaps something like
```python
class sin(...)
    def _eval_is_bounded(self, symbol):
        # Some facts may depend on `symbol`, but if the argument is real, this is bounded anyway
        if self.args[0].is_real:
            return True
        # And now the complicated part should come...
        # Whether args have symbol or not may affect the result, and how exactly the symbol affects args, etc...

    def _eval_is_odd_function(self, symbol):
        # As `sin` itself is `odd`, this is a composition of `odd` on `self.args[0]`
        # Of course, I don't really want to call a private method of the arg; this is only an illustration
        # Note that by default, every symbol is odd with respect to itself
        if self.args[0]._eval_is_odd_function(symbol):
            # It is even in this case
            return False
        # Note that by default, every symbol is even with respect to any other symbol
        if self.args[0]._eval_is_even_function(symbol):
            return True
```

**Comment by pelegm** [comment]
> One thing I would caution about is to be very precise in your definitions of the assumptions.

Agree. Do you have any idea where can I define all of these? Perhaps I should create a new text/rst file somewhere (in `sympy/assumptions`?), where the definitions and the abstract "facts" would slowly add up?

**Comment by pelegm** [comment]
> A more common term for your "additive" is "linear".

Actually, I was not careful before. When I wrote `linear` I actually meant `linear function`, not `linear map`. That is, I wasn't intending to require `f(0)=0`, so, for example, `ax+b` is linear (but not additive if `b != 0`).
